pkg/util/ipsec/ipsec_telemetry.go

```diff
@@ -0,0 +1,50 @@
+package ipsec
+
+import (
+	"sync"
+
+	operv1 "github.com/openshift/api/operator/v1"
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+var (
+	registerMetrics sync.Once
+	ipsecStateGauge *metrics.GaugeVec
+
+	recordedMode, recordedFlavor string = "", ""
+)
+
+// RegisterIpsecMetrics in the global registry
+func RegisterIpsecMetrics() {
+	registerMetrics.Do(func() {
+		ipsecStateGauge = metrics.NewGaugeVec(&metrics.GaugeOpts{
+			Name: "cluster_network_ipsec_state",
+			Help: "A metric with a constant '1' value labeled by the latest ipsecMode of the cluster",
+		}, []string{"mode", "apiFlavor"})
+		legacyregistry.MustRegister(ipsecStateGauge)
+	})
+}
+
+func UpdateIpsecTelemetry(ipsecConfig *operv1.IPsecConfig) {
+	var flavor, mode string
+	klog.V(5).Infof("IPsec Telemetry: %v", ipsecConfig)
+	RegisterIpsecMetrics()
+	if ipsecConfig == nil {
+		mode = string(operv1.IPsecModeDisabled)
+		flavor = "4.14"
+	} else if ipsecConfig.Mode == "" {
+		mode = string(operv1.IPsecModeFull)
+		flavor = "4.14"
+	} else {
+		mode = string(ipsecConfig.Mode)
+		flavor = "4.15"
+	}
+	if mode != recordedMode || flavor != recordedFlavor {
+		klog.V(5).Infof("IPsec Telemetry: (%s, %s), writing to %v", mode, flavor, ipsecStateGauge)
+		ipsecStateGauge.WithLabelValues(mode, flavor).Set(1)
```

Comment:
I'm neither an expert on instrumentation nor sure whether the following is intentional.
`WithLabelValues` will create a new time series if you call it with a never seen before set of label values. So if the expectation here is to always have a single timeseries, you'll need to reset the vector before set. Otherwise you'll end up with multiple series if those label values change.

Reply:
@jan--f Thanks - I am not a Prometheus expert and I wondered how to keep a single value (It was not critical,  since in most cases we will have 1 or 2 values, but what I want is always a single value).

Reply:
Right, this is unfortunately a discouraged but widespread practice. Prometheus best-practice would recommend to always emit all possible label dimensions and set 0 for the ones not observed.
That can however be expensive. For larger metric sets like this a custom collector is usually used, like in https://github.com/kubernetes/kubernetes/blob/20d0ab7ae808aaddb1556c3c38ca0607663c50ac/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L299

I suppose this metric is expected to rarely change in a running cluster? So the reset way is probably just fine here.

Reply:
As @jan--f  said, the preferred option would be to expose all label values with a value of 1 for the mode being enabled:

```
cluster_network_ipsec_state{mode="disabled"} 0
cluster_network_ipsec_state{mode="external"} 1
cluster_network_ipsec_state{mode="full"} 0
```

Using Reset() should work fine though it's more error-prone. The other option if you want to keep only 1 series would be to use the custom collector pattern (https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics).

Reply:
@jan--f - this metric is indeed expected to change very rarely in production clusters, once you enable ipsec you rarely disable it (assuming you had a business reason to enable it, it doesn't go away).

Reply:
@simonpasquier - for a metric as simple as this which rarely changes, can I leave the current implementation? I believe the option you suggest will just add more lines of code for something really simple and will make the code cumbersome.

Reply:
I'm not sure that the code would be more complex with a custom collector. It's your call since you maintain that piece of code ;)